storm/src/main/storm/mesos/resources/ScalarResource.java

```diff
@@ -0,0 +1,165 @@
+    ScalarResourceEntry availableResource = availableResourcesByReservationType.get(reservationType);
+    List<ResourceEntry> reservedResources = new ArrayList<>();
+
+    if (scalarResourceEntry.getValue() < availableResource.getValue()) {
+      availableResourcesByReservationType.put(reservationType, availableResource.remove(scalarResourceEntry));
+      totalAvailableResource -= scalarResourceEntry.getValue();
+      reservedResources.add(new ScalarResourceEntry(scalarResourceEntry.getReservationType(), scalarResourceEntry.getValue()));
+      return reservedResources;
+    }
+    String message = String.format("ResourceType '%s' of reservationType '%s' is not available. Requested value: %s Available: %s",
+                                   resourceType, reservationType.toString(), scalarResourceEntry.getValue(), availableResourcesByReservationType.get(reservationType));
+    throw new ResourceNotAvailabeException(message);
+  }
+
+  public String toString() {
+    List<String> availableResourcesByResourceTypeList = new ArrayList<>();
+    for (Map.Entry<ReservationType, ScalarResourceEntry> entry: availableResourcesByReservationType.entrySet()) {
+      availableResourcesByResourceTypeList.add(String.format("%s : %s", entry.getKey(), entry.getValue().getValue()));
+    }
+    String tmp = StringUtils.join(", ", availableResourcesByResourceTypeList);
+    return String.format("Resource %s - Total available : %f Total available by reservation type : [ %s ]", resourceType.toString(), totalAvailableResource, tmp);
+  }
+
+  private List<ResourceEntry> removeAndGet(ScalarResourceEntry scalarResourceEntry, Collection<ReservationType> reservationTypesListByPriority) throws ResourceNotAvailabeException {
+    Double requiredValue = scalarResourceEntry.getValue();
+    List<ResourceEntry> reservedResources = new ArrayList<>();
+
+    if (requiredValue > totalAvailableResource) {
+      String message = String.format("ResourceType '%s' is not available. Requested value: %s Available: %s",
+                                     resourceType, requiredValue, totalAvailableResource);
+      throw new ResourceNotAvailabeException(message);
+    }
+
+    for (ReservationType reservationType : reservationTypesListByPriority) {
+      ScalarResourceEntry availableResource = availableResourcesByReservationType.get(reservationType);
+      Double availableResourceValue = availableResource.getValue();
+
+      if (availableResourceValue >= requiredValue) {
+        availableResource.remove(new ScalarResourceEntry(requiredValue));
+        totalAvailableResource -= requiredValue;
+        reservedResources.add(new ScalarResourceEntry(reservationType, requiredValue));
+        return reservedResources;
+      } else if (availableResourceValue > 0) {
+        availableResourcesByReservationType.put(reservationType, new ScalarResourceEntry(reservationType, 0.0));
```

Comment:
The way I'm reading it is that the logic here flows as follows
```
if (requested resources fit entirely into available resources)
    remove the requested resources form the available resources
    reduce the total available resources by the requested resources
    reserve the requested resources so that we can use them
    return the reserved resources
else if (there are still available resources, but not all requested resources are available)
    add an "empty" resource to the list of available resources for this reservation type
    reduce the total available resources by the amount of resources we are using here
    reserve the available resource so that we can use them later
```
I'm not positive I understand why we wouldn't just remove the entire resource by calling `availableResource.remove(new ScalarResourceEntry(availableResourceValue))` (as is similar to what is done in the `if`) and why the `else if` requires that we pay special attention to the `reservationType` (where the `if` seems to ignore the `reservationType` until it goes to reserve the resources).